### ISSUE TITLE
bodyfile generated by MFTECmd contains incorrect TSK metadata addresses

### ISSUE THREAD

**Issue by joachimmetz** [EricZimmerman/MFTECmd#5]
Related to https://github.com/sleuthkit/sleuthkit/issues/1809, bodyfile sent to Plaso project

```
0|c:/testdir16/hardlink15|38-128-5|r/rrwxrwxrwx|0|0|14|1580550526|1580550526|1580550526|1580550526
0|c:/testdir16/hardlink15 ($FILE_NAME)|38-48-19|r/rrwxrwxrwx|0|0|14|1580550526|1580550526|1580550526|1580550526
0|c:/testdir3/hardlink2|38-128-5|r/rrwxrwxrwx|0|0|14|1580550526|1580550526|1580550526|1580550526
0|c:/testdir3/hardlink2 ($FILE_NAME)|38-48-0|r/rrwxrwxrwx|0|0|14|1580550526|1580550526|1580550526|1580550526
0|c:/testdir4/hardlink3|38-128-5|r/rrwxrwxrwx|0|0|14|1580550526|1580550526|1580550526|1580550526
0|c:/testdir4/hardlink3 ($FILE_NAME)|38-48-1|r/rrwxrwxrwx|0|0|14|1580550526|1580550526|1580550526|1580550526
0|c:/testdir5/hardlink4|38-128-5|r/rrwxrwxrwx|0|0|14|1580550526|1580550526|1580550526|1580550526
0|c:/testdir5/hardlink4 ($FILE_NAME)|38-48-2|r/rrwxrwxrwx|0|0|14|1580550526|1580550526|1580550526|1580550526
0|c:/testdir6/hardlink5|38-128-5|r/rrwxrwxrwx|0|0|14|1580550526|1580550526|1580550526|1580550526
0|c:/testdir6/hardlink5 ($FILE_NAME)|38-48-3|r/rrwxrwxrwx|0|0|14|1580550526|1580550526|1580550526|1580550526
0|c:/testdir7/hardlink6|38-128-5|r/rrwxrwxrwx|0|0|14|1580550526|1580550526|1580550526|1580550526
```

```
icat -o 128 ntfs_file_name_list.vhd 38-48-1 | hexdump -Cv
Attribute not found in file (tsk_fs_attrlist_get_id: Attribute 48-1 not found)

icat -o 128 ntfs_file_name_list.vhd 38-48-3 | hexdump -Cv
00000000  27 00 00 00 00 00 01 00  7a 39 d8 cb e4 d8 d5 01  |'.......z9......|
00000010  7a 39 d8 cb e4 d8 d5 01  f1 37 d9 cb e4 d8 d5 01  |z9.......7......|
00000020  7a 39 d8 cb e4 d8 d5 01  10 00 00 00 00 00 00 00  |z9..............|
00000030  0e 00 00 00 00 00 00 00  20 00 00 00 00 00 00 00  |........ .......|
00000040  09 00 68 00 61 00 72 00  64 00 6c 00 69 00 6e 00  |..h.a.r.d.l.i.n.|
00000050  6b 00 31 00                                       |k.1.|
```

To cite your earlier claims "They say speed is fine, but accuracy is final. MFTECmd fits the bill in both regards." yeah you're not convincing me.

**Comment by EricZimmerman** [comment]
Nice dig. I'm glad all your software works perfectly!

I don't care what tsk says. Tsk has been wrong many times (yet you seem to assume it is correct here. It may very well be, but who knows without looking at the file records.). I found tsk issues when I was writing mfrecmd. 

You can get me the mft file or go look at the file records (and any extension records since this seems to involve hard links) yourself to see what things look like. 

I still stand by my statement. I never said mfrecmd is perfect, but you do you. No software is perfect as you well know

**Comment by EricZimmerman** [comment]
also, how nice of you to add comments to an old issue where no one can reply at https://github.com/log2timeline/plaso/issues/1466#issuecomment-666193309

the thing is, this is OPEN SOURCE software. if people want all that precision, they can have it. 

if i recall, i emulated what fls does. if someone wants to see subsecond info, dump to csv or ask and i can see what i can do, but you seem to be the only one complaining.

**Comment by joachimmetz** [comment]
You are not getting the point. This is people reporting to the Plaso project, based on **your recommendation**. They used one of the test images I'd created and tested TSK wtih. 

> also, how nice of you to add comments to an old issue where no one can reply at log2timeline/plaso#1466 (comment)

Yes to prevent you from providing more bad recommendations and off topic promotion of MFTECmd.

> I don't care what tsk says.

I don't understand your point. You are using the part of the body file that is called the "TSK metadata address" https://wiki.sleuthkit.org/index.php?title=Metadata_Address. You either implement it as TSK, which is wrong https://github.com/sleuthkit/sleuthkit/issues/1809, or you don't. 

The current implementation by MFTECmd is not the same as TSK nor does it correspond with the attribute list values (see the output of istat). So it is not accurate nor has it value since it does not allow to correspond back to the original `$FILE_NAME` attribute.

> if i recall, i emulated what fls does. if someone wants to see subsecond info, dump to csv or ask and i can see what i can do, but you seem to be the only one complaining.

You are the one that was indicating that "accuracy is final." and boasting that "MFTECmd fits the bill in both regards." and now you play it off as "No software is perfect as you well know".

I know perfectly well no software is perfect, but I do not go boasting advising on random project that my tools are.

**Comment by joachimmetz** [comment]
> if i recall, i emulated what fls does. if someone wants to see subsecond info, dump to csv or ask and i can see what i can do, but you seem to be the only one complaining.

**You recommended** people to use MFTECmd to create a bodyfile and ingest that into Plaso, I'm now telling you that the output MFTECmd creates is **not accurate**. Do what you want to do with that information.

**Comment by EricZimmerman** [comment]
can you share your test image, or at least the MFT? that would let me look into it to see how the FILE records look and then compare it to bodyfile output. if not, i cant do much with this.

i never strived to make body file 100% accurate with tsk. i want to reflect what is in the FILE records (base + extension records) from a name and timestamp perspective. i am not a fan of bodyfile for many reasons but people wanted it added, so I did.

mftecmd works well in the vast majority of cases. i still havent ever said it is perfect, but you again imply i did in your last statement ("boasting advising on random project").

istat was wrong in my previous testing, so again, if you have a specific, concrete example (i.e. the $MFT file or test image), lets discuss it. if not, not much i can do here without a test case

i am not going to argue with you about who said what was perfect or not. I can find cases where lots of software does not get things right.

**TL;DR
if you have a specific, concrete example (i.e. the $MFT file or test image), lets discuss it. if not, not much i can do here without a test case. Please let me know**

**Comment by joachimmetz** [comment]
From the linked TSK issue
```
Based on the ntfs_file_name_list.vhd test image generated by https://github.com/dfirlabs/ntfs-specimens/blob/master/generate-specimens-behavior.bat this does not appear to be true. In the output below different attributes have duplicate TSK metadata addresses.
```

**Comment by joachimmetz** [comment]
> mftecmd works well in the vast majority of cases. i still havent ever said it is perfect, but you again imply i did in your last statement ("boasting advising on random project").

I'm citing you directly from https://github.com/log2timeline/plaso/issues/1466#issuecomment-424840036

**Comment by EricZimmerman** [comment]
the word "perfect" does not appear anywhere in that post.

also, that .bat file does not create any vhd named 'ntfs_file_name_list.vhd'

which one to use?

ntfs-scenario3.1.vhd
ntfs-scenario3.2.vhd
ntfs_path_hint.vhd
ntfs_scenario1.1.vhd
ntfs_scenario1.2.vhd
ntfs_scenario1.3.vhd
ntfs_scenario1.4.vhd
ntfs_scenario1.5.vhd
ntfs_scenario1.6.vhd

**Comment by joachimmetz** [comment]
> the word "perfect" does not appear anywhere in that post.

? you are the one that brought up perfect not me in this thread, not me.

> also, that .bat file does not create any vhd named 'ntfs_file_name_list.vhd'

Things have progressed since that issue was filed it is called ntfs-scenario3.1.vhd now

**Comment by EricZimmerman** [comment]
ill take a look at 3.1

**Comment by EricZimmerman** [comment]
so, what specifically is the problem?

this is in MFTECmd body output:

0|c:/testdir2/hardlink1 ($FILE_NAME)|37-48-3|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157

this shows the hardlink1 info:

icat -o 128 .\ntfs-scenario3.1.vhd 37-48-3

This shows everything about entry 37:

.\MFTECmd.exe -f 'C:\Temp\aaa\$MFT' --de 37

![image](https://user-images.githubusercontent.com/4265254/88936037-6449db80-d250-11ea-85e1-ae51cbd93ff5.png)


here is more stuff from body file from that MFT:

0|c:/testdir2/hardlink1|37-128-5|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir2/hardlink1 ($FILE_NAME)|37-48-3|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir16/hardlink15|37-128-5|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir16/hardlink15 ($FILE_NAME)|37-48-19|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir3/hardlink2|37-128-5|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir3/hardlink2 ($FILE_NAME)|37-48-0|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir4/hardlink3|37-128-5|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir4/hardlink3 ($FILE_NAME)|37-48-1|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir5/hardlink4|37-128-5|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir5/hardlink4 ($FILE_NAME)|37-48-2|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir6/hardlink5|37-128-5|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir6/hardlink5 ($FILE_NAME)|37-48-3|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir7/hardlink6|37-128-5|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir7/hardlink6 ($FILE_NAME)|37-48-4|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir8/hardlink7|37-128-5|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir8/hardlink7 ($FILE_NAME)|37-48-0|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir9/hardlink8|37-128-5|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir9/hardlink8 ($FILE_NAME)|37-48-1|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir10/hardlink9|37-128-5|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir10/hardlink9 ($FILE_NAME)|37-48-2|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir11/hardlink10|37-128-5|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157

maybe someone sent you in body file from an older version of MFTECmd.

if you see something broken still, please let me know. 

i would have used fls to generate body file, but it was not happy (i imaged the vhd to a raw image):

fls -f ntfs -r -m c: C:\temp\ntfs-scenario3.1.001
Invalid magic value (Not a NTFS file system (invalid sector size 190)))

if you can elaborate on exactly where things are "not accurate and lacks precision" for this body file, and mft, i would love to address the issue

[aaa.zip](https://github.com/EricZimmerman/MFTECmd/files/5001872/aaa.zip)

**Comment by joachimmetz** [comment]
> so, what specifically is the problem?

The problem is that in the bodyfile sent to me the TSK metadata address did not correspond with that of the image. Nor does it in your output. E.g. `37-48-3` is referenced twice, that means at least one of them does not refer back to the original `$FILE_NAME` attribute. 

```
0|c:/testdir6/hardlink5 ($FILE_NAME)|37-48-3|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
0|c:/testdir2/hardlink1 ($FILE_NAME)|37-48-3|r/rrwxrwxrwx|0|0|14|1596118157|1596118157|1596118157|1596118157
```

> maybe someone sent you in body file from an older version of MFTECmd.

based on your output the issue still to pesist.

> i would have used fls to generate body file, but it was not happy (i imaged the vhd to a raw image):

Why? the fixed vhd is a raw image with an additional footer

> fls -f ntfs -r -m c: C:\temp\ntfs-scenario3.1.001

It contains a full partition image, try mmls

**Comment by EricZimmerman** [comment]
i totally agree with you on the precision, however, i modeled what TSK was doing when it generated bodyfile. i much prefer the subsecond precision too.

as to the other attribute # thing, i am looking at that now.

**Comment by EricZimmerman** [comment]
ok i see where the discrepancy is. i dont think its an issue and i would argue tsk is not doing things properly here

The data for the hardlinks are in extension records. i have expanded them here.

Notice the attribute numbers do not just keep going up, but reset in each FILE record. this is why you are seeing them duplicated.

![image](https://user-images.githubusercontent.com/4265254/88949456-40db5c80-d261-11ea-9d0f-58bb1cdc8fa9.png)

here is the base record:

![image](https://user-images.githubusercontent.com/4265254/88949786-b2b3a600-d261-11ea-94a7-6852226db20e.png)

i gather all extension records and then include them as output related to the base record. this is why you are seeing what you are seeing.

How is tsk determining which attribute # it is and being shown? maybe they are just incrementing a counter or something, but thats not the true attribute # as it exists in the file record that i can tell.

now perhaps i can adjust the stuff in the extension records to point  to the entry number they are stored in, but i wanted to tie all the attributes from the main FILE record together. This is why things were done they way they were

i have no idea how tsk is doing things, so it is a discrepancy, but i am using the actual attribute # as stored in the attribute itself vs some computed value. if there is a way that NTFS is presenting that that can be gleaned from the MFT, please let me know and i will gladly update things. 

thoughts?

**Comment by joachimmetz** [comment]
>  dont think its an issue and i would argue tsk is not doing things properly here

TSK is definitely wrong here: see https://github.com/sleuthkit/sleuthkit/issues/1809

I recommend removing the `-48-3` and keeping away from trying to represent poorly documented and inconsistent TSK internals.

> Notice the attribute numbers do not just keep going up, but reset in each FILE record. this is why you are seeing them duplicated.

I understand, however from the previous TSK issue "TSK overrides those ID values though and assigns new ones so that each attribute is unique.", which is neither the true for the current TSK or MFTECmd implementation.

> How is tsk determining which attribute # it is and being shown? 

Unclear, but I can imagine this to be depending on the order of which `$ATTRIBUTE_LIST` data MFT entry is read first.

> i wanted to tie all the attributes from the main FILE record together.

Based on the TSK documentation I would say the intent was to make them all unique, which would require to use the `$ATTRIBUTE_LIST` data MFT entry and attribute index number within the MFT entry, however you would lose the relation to the base MFT entry. My take on it is to only use the MFT entry number and not shoehorn TSK internals into a data format used for interchange between tools.

**Comment by EricZimmerman** [comment]
i parse loose MFTs, so the attribute list, when non-resident, would break that approach using the list. thats why i went with finding and connecting up the extension records.

**Comment by joachimmetz** [comment]
> i parse loose MFTs, so the attribute list, when non-resident, would break that approach using the list. thats why i went with finding and connecting up the extension records.

Unclear to me which part of my reply you are specifically referring to.

However I can image the approach you use.
I was facing a similar problem and that is why I created the test image.
And realized that what TSK is doing cannot be portable.
However looking closer at it, it does not even appear to be working as intended.

**Comment by EricZimmerman** [comment]
so this boils down to tsk not working well/consistently. =(

i will tweak the body file format mfecmd is generating to make it clear which FILE record the data is coming from i suppose, to at least lead people to where its coming from

as to the attribute list, i was just adding a note about how/why i did it the way that I do. if tools parse the whole image they would have the non-resident lists and its less of a concern, but i wanted to handle loose MFTs, so I did it the way i did:

https://github.com/EricZimmerman/MFT/blob/master/MFT/Mft.cs#L217

**Comment by joachimmetz** [comment]
> so this boils down to tsk not working well/consistently. =(

This boils down MFTECmd copying behavior from TSK that is not working well/consistently. And that it is **not** a good recommendation to use this information in bodyfiles that are processed by tools that do expect this information to be accurate by you on log2timeline/plaso#1466 and various SANS teachers, just because they think this is a better solution without actually looking into it.

If this was just a TSK issue, we would not have to have this chat on this issue tracker.

**Comment by EricZimmerman** [comment]
my bad then for trying to emulate tsk and assuming things on how it works. i didnt want to do bodyfile at all because i think its a bit of a trainwreck myself, but people wanted it, so i added it.

everything but that part of the output seems to be in order tho.
